R/render-cached-plot.R

```diff
@@ -0,0 +1,601 @@
+  # necessary in some cases.
+  drawReactiveTrigger <- reactiveVal(0)
+
+  # Calls drawPlot, invoking the user-provided `func` (which may or may not
+  # return a promise). The idea is that the (cached) return value from this
+  # reactive can be used for varying width/heights, as it includes the
+  # displaylist, which is resolution independent.
+  drawReactive <- reactive(label = "plotObj", {
+    hybrid_chain(
+      {
+        # Get width/height, but don't depend on them.
+        isolate({
+          # The first execution will have NULL width/height, because they haven't
+          # yet been retrieved from clientData.
+          req(fitDims$width, fitDims$height, cancelOutput = TRUE)
+        })
+
+        drawReactiveTrigger()
+        # cat("drawReactive()\n")
+
+        cacheKey()
+      },
+      function(cacheKeyResult) {
+        # Get width/height, but don't depend on them.
+        isolate({
+          width  <- fitDims$width
+          height <- fitDims$height
+        })
+
+        pixelratio <- session$clientData$pixelratio %OR% 1
+
+        key <- digest::digest(list(outputName, cacheKeyResult, width, height, res, pixelratio), "sha256")
+
+        cached_value <- cache$get(key)
+
+        if (!is.key_missing(cached_value)) {
+          # cat("drawReactive(): cached\n")
+          # This will NOT include the displaylist.
+          return(cached_value)
+
+        } else {
+          hybrid_chain(
+            {
+              # cat("drawReactive(): drawPlot()\n")
+              # This will include the displaylist.
+              do.call("drawPlot", c(
+                list(
+                  name = outputName,
+                  session = session,
+                  func = isolatedFunc,
+                  width = width,
+                  height = height,
+                  pixelratio = pixelratio,
+                  res = res
+                ),
+                args
+              ))
+            },
+            function(result) {
+              # Cache a copy of the result. In the case of the MemoryCache,
+              # the result is simply stored in memory and can be restored just
+              # fine. For other types of cache (like DiskCache) where the
+              # object must be serialized before saving, there's a catch: the
+              # recorded displaylist for the plot can't be serialized and
+              # restored properly within the same R session, so we NULL it out
+              # before saving. (The PNG can of course be saved and restored
+              # just fine.) Note that this was fixed in revision 74506
+              # (2e6c669), and should be in R 3.6, but we need it to work on
+              # older versions. Perhaps in the future we could do a version
+              # check and change caching behavior based on that.
+              result_copy <- result
+              if (!inherits(cache, "MemoryCache")) {
+                result_copy$recordedPlot <- NULL
```

Comment:
Let's not include recordedPlot in the cached value. The cache key includes the size, so there isn't a case where you'd hit on the displaylist but not on a png.